src/App/Datums.cpp

```diff
@@ -0,0 +1,297 @@
+    }
+    setupData[] = {
+        // clang-format off
+        {App::Line::getClassTypeId(),  AxisRoles[0],  tr("X-axis"),   Base::Rotation()},
+        {App::Line::getClassTypeId(),  AxisRoles[1],  tr("Y-axis"),   Base::Rotation(Base::Vector3d(1,1,1), M_PI * 2 / 3)},
+        {App::Line::getClassTypeId(),  AxisRoles[2],  tr("Z-axis"),   Base::Rotation(Base::Vector3d(1,-1,1), M_PI * 2 / 3)},
+        {App::Plane::getClassTypeId(), PlaneRoles[0], tr("XY-plane"), Base::Rotation()},
+        {App::Plane::getClassTypeId(), PlaneRoles[1], tr("XZ-plane"), Base::Rotation(1.0, 0.0, 0.0, 1.0)},
+        {App::Plane::getClassTypeId(), PlaneRoles[2], tr("YZ-plane"), Base::Rotation(Base::Vector3d(1,1,1), M_PI * 2 / 3)},
+        {App::Point::getClassTypeId(), PointRoles[0], tr("Origin"),   Base::Rotation()}
+        // clang-format on
+    };
+
+    App::Document* doc = getDocument();
+
+    std::vector<App::DocumentObject*> links;
+    for (auto data : setupData) {
+        std::string objName = doc->getUniqueObjectName(data.role);
+        App::DocumentObject* featureObj = doc->addObject(data.type.getName(), objName.c_str());
+
+        assert(featureObj && featureObj->isDerivedFrom(App::DatumElement::getClassTypeId()));
+
+        QByteArray byteArray = data.label.toUtf8();
+        featureObj->Label.setValue(byteArray.constData());
+
+        App::DatumElement* feature = static_cast <App::DatumElement*> (featureObj);
+        feature->Placement.setValue(Base::Placement(Base::Vector3d(), data.rot));
+        feature->Role.setValue(data.role);
+
+        feature->Placement.setStatus(Property::Hidden, true);
+
+        links.push_back(feature);
+    }
+
+    OriginFeatures.setValues(links);
+}
+
+void LocalCoordinateSystem::unsetupObject() {
+    const auto& objsLnk = OriginFeatures.getValues();
+    // Copy to set to assert we won't call methode more then one time for each object
+    std::set<App::DocumentObject*> objs(objsLnk.begin(), objsLnk.end());
+    // Remove all controlled objects
+    for (auto obj : objs) {
+        // Check that previous deletes wasn't inderectly removed one of our objects
+        const auto& objsLnk = OriginFeatures.getValues();
+        if (std::find(objsLnk.begin(), objsLnk.end(), obj) != objsLnk.end()) {
+            if (!obj->isRemoving()) {
+                obj->getDocument()->removeObject(obj->getNameInDocument());
+            }
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+LocalCoordinateSystem::LCSExtension::LCSExtension(LocalCoordinateSystem * obj)
+    : obj(obj)
+{
+    Group.setStatus(Property::Transient, true);
+}
+
+void LocalCoordinateSystem::LCSExtension::initExtension(ExtensionContainer * obj) {
+    App::GroupExtension::initExtension(obj);
+}
+
+bool LocalCoordinateSystem::LCSExtension::extensionGetSubObject(DocumentObject * &ret, const char* subname,
+    PyObject * *pyobj, Base::Matrix4D * mat, bool, int depth) const {
```

Comment:
Pretty strange formatting with `PyObject * *pyobj`, this whole file is new and so should be formatted using clang-format.

Reply:
The file is new but the code is old so it carries the previous formatting. I'll try to get it fixed.